docker/configure_workers_and_start.py

```diff
@@ -490,46 +635,171 @@ def generate_worker_files(
+                    + worker_type
+                    + ". Please fix!"
+                )
+
+            # Merge worker type template configuration data.
+            worker_config = merge_worker_template_configs(
+                worker_config, copy_of_template_config
+            )
+
+            # Check if there is to many of a worker there can only be one of.
+            # Will error and stop if it is a problem, e.g. 'background_worker'.
+            if worker in worker_type_shard_counter:
+                if not is_sharding_allowed_for_worker_type(worker):
+                    error(
+                        "There can be only a single worker with "
+                        + worker
+                        + " type. Please recount and remove."
+                    )
+            # Not in shard counter, add it. Don't need return value
+            increment_counter(worker_type_shard_counter, worker)
+
+        # Name workers by their type or requested name concatenated with an
+        # incrementing number. e.g. federation_reader1 or event_creator+event_persister1
+        if requested_worker_name:
+            worker_base_name = requested_worker_name
+            # It'll be useful to have this in the log in case it's a complex of many
+            # workers merged together. Note for Complement: it would only be seen in the
+            # logs for blueprint construction(which are not collected).
+            log(
+                "Worker name request found: "
+                + requested_worker_name
+                + ", for: "
+                + str(workers_to_combo_list)
+            )
+
         else:
-            error(worker_type + " is an unknown worker type! Please fix!")
+            # The worker name will be the worker_type, however if spaces exist
+            # between concatenated worker_types and the "+" because of readability,
+            # it will error on startup. Recombine worker_types without spaces and log.
+            if len(workers_to_combo_list) > 1:
+                worker_base_name = "+".join(workers_to_combo_list)
+                if worker_base_name != worker_type:
+                    log(
+                        "Default worker name would have contained spaces, which is not "
+                        "allowed(" + worker_type + "). Reformed name to not contain "
+                        "spaces: " + worker_base_name
+                    )
+            else:
+                worker_base_name = worker_type
+        # This counter is used for names with broader worker type definitions.
+        new_worker_count = increment_counter(worker_type_counter, worker_type)
+        worker_name = worker_base_name + str(new_worker_count)
+
+        # Now that the worker name is settled, check this name isn't used for a
+        # different worker_type. If it's not allowed, will error and stop. If no issues,
+        # it will be added to the counter.
+        if is_name_allowed_for_worker(
```

Comment:
I'm not sure why this isn't just e.g.

```python
existing_worker = worker_name_type_counter.get(worker_base_name)
if existing_worker:
    error(f"Can not use {worker_name} for {worker_type}. It is already in use by {existing_worker}.")
worker_name_type_counter[worker_base_name] = worker_type
```
(I'd probably change the name of `worker_name_type_counter` — counter seems to be a misnomer?)

Reply:
```
'bob=pusher:2'
worker_base_name=bob
worker_name=bob2
```
If I followed what you coded there, this would see that bob had been registered and reject it's further use. We don't want that, we want to see if bob has already been registered as a pusher because if he's not then naming this worker bob would be Bad News.

(I will rename this counter as you are definitely right about the misnomer)

Reply:
This has become an if statement and the original function has been removed. Hopefully this will be clearer in the context.